### DISCUSSION TITLE
OpenSSL migration | Replacement for APIs used in RSA PKCS1OAEPPadding MGF1 SHA256/384/512 related custom functions

### DISCUSSION THREAD

**Discussion by gsharma-ad** [openssl/openssl#21575] [Q&A] [answered]
**OpenSSL 3.0.8**

We are fixing OpenSSL deprecated warnings with OpenSSL 3.0.8.
During this we need to replace below APIs used in custom encrypt\decrypt for RSA PKCS1OAEPPadding SHA256/384/512 functions-:
```
warning: ‘const RSA_METHOD* RSA_PKCS1_OpenSSL()’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:305)
warning: ‘int (* RSA_meth_get_mod_exp(const RSA_METHOD*))(BIGNUM*, const BIGNUM*, RSA*, BN_CTX*)’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:525)
warning: ‘int (* RSA_meth_get_bn_mod_exp(const RSA_METHOD*))(BIGNUM*, const BIGNUM*, const BIGNUM*, const BIGNUM*, BN_CTX*, BN_MONT_CTX*)’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:533)
warning: ‘int PKCS1_MGF1(unsigned char*, long int, const unsigned char*, long int, const EVP_MD*)’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:389) 
```

We are unable to find replacement for these APIs. As per migration guide, we shall use **provider** but we are not able to conclude from provider page.

To us it seems that we need to continue using these APIs until OpenSSL provides any replacement for these APIs.
Please suggest.

Excuse me if I am asking very basic question.

**Comment by mattcaswell** [comment, marked as answer]
> warning: ‘const RSA_METHOD* RSA_PKCS1_OpenSSL()’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:305)

RSA_METHOD is deprecated so anything touching it is also deprecated. In 3.x you instead implement a provider based implementation of RSA.

> warning: ‘int (* RSA_meth_get_mod_exp(const RSA_METHOD*))(BIGNUM*, const BIGNUM*, RSA*, BN_CTX*)’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:525)
> warning: ‘int (* RSA_meth_get_bn_mod_exp(const RSA_METHOD*))(BIGNUM*, const BIGNUM*, const BIGNUM*, const BIGNUM*, BN_CTX*, BN_MONT_CTX*)’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:533)

Unfortunately, the provider based interface does not offer replacements for these. There are currently no plans to do this.

> warning: ‘int PKCS1_MGF1(unsigned char*, long int, const unsigned char*, long int, const EVP_MD*)’ is deprecated (declared at ../../../../../build/Linux_RedHat_6.9_64/include/openssl/rsa.h:389) 

Not sure about this one.

**Comment by t8m** [comment]
What is the PKCS1_MGF1() function used for in your application? The RSA OAEP padding is done internally by the RSA encryption decryption if the OAEP padding mode is set.

There are no plans to provide PKCS1_MGF1() function standalone via the provider API as this was seen as a low-level operation. However I assume it could be added as a KDF operation.

**Reply by gsharma-ad** [reply]
**RSA OAEP padding is done internally**: For us this working fine for PKCS1OAEPPadding only.

We are using PKCS1_MGF1() to generate some masking on MGF1 hash for PKCS1OAEPPaddingSHA256/384/512, pretty similar to OpenSSL API  **RSA_padding_check_PKCS1_OAEP_mgf1**.
